src/Kestrel.Core/Internal/Http/OutputProducer.cs

```diff
@@ -148,6 +149,47 @@ public void Abort(Exception error)
             return FlushAsync(writableBuffer, cancellationToken);
         }
 
+        public async Task WriteAsync(
+            Stream stream,
+            long length,
+            CancellationToken cancellationToken)
+        {
+
+            var writableBuffer = default(WritableBuffer);
+            var remaining = length;
+            int count;
+            do
+            {
+                lock (_contextLock)
```

Comment:
@halter73 would this locking be enough?

Reply:
Given our current locking strategy which is to ensure that the contents of concurrent calls to Response.Body.Write and similar don't interleave, then no.

I'm not in love with the strictness of this strictness of this strategy, but at least now it should be low contention unlike in the 1.x versions of Kestrel where the _contextLock would be shared with the consumer on the libuv thread.

Reply:
We might have to change to a semaphore

Reply:
Semaphore is allocately. Gate that works for sync and async?

Reply:
This semaphore, is that something that a user of `IHttpSendFileFeature` must manually lock/unlock to ensure the entire response doesn't interleave with other output?

Reply:
Today we prevent overlapping writes by locking. SendFile is basically a write.

Reply:
Would be automatic

Reply:
If you have some code like:
```
WriteHeaders
SendFile
```
This unlocks in-between. Can http2 interleave responses in a way that this would be an issue?

Reply:
What is WriteHeaders?

> This unlocks in-between. Can http2 interleave responses in a way that this would be an issue?

Nope.

Reply:
> What is WriteHeaders?

What ever code writing the headers before sendfile is writing the content.

> Nope.

In general, why don't responses interleaved? Does a response get packed as a "WritableBuffer"? Or is there some other lock?

Reply:
> What ever code is writing the headers before sendfile is writing the content.

Unlocking in between those calls is fine. We're just making sure the buffer isn't corrupted (i.e. 2 calls don't write to the WritableBuffer in parallel). 

> In general, why don't responses interleave? Does a response get packed as a "WritableBuffer"? Or is there some other lock?

I'm not sure what you mean?

Reply:
For a response, something is 'grouping' the data that needs to be sent together (e.g on Flush). Then this 'group' is sent as a a whole (so no data from other responses can interleave).
In general: How is this implemented? Is this grouping done by `WritableBuffer` and is `_contextLock` responsible for appending a `WritableBuffer` to the output stream? Or something else?
Here specifically: How does `SendFileAsync` get 'grouped' with the response headers?

Reply:
I think you are going too deep; it doesn't compile yet 😉 

Though it should check the headers and trigger them prior to sending the file data? (i.e. they are normal triggered by a `Response.Body.Write` which this bypasses)

Transport level send file wouldn't be able to do that; so it should intercept even if the transport supported `SendFile` for prep-work?

If `ContentLength` is not set it should also chunk? Not sure how large a single chunk can be (i.e. whole file) as it shouldn't chunk inside the file data, or is altering `SendFile`'s expected behaviour too much (i.e. just top and tail)?

Reply:
> Though it should check the headers and trigger them prior to sending the file data?

Yes, if headers are not sent I think they should be at that point.

> If ContentLength is not set it should also chunk? Not sure how large a single chunk can be (i.e. whole file) as it shouldn't chunk inside the file data, or is altering SendFile's expected behaviour too much (i.e. just top and tail)?

I think that would be appropriate. So when `long? count` is `null`, it should be set it to `File.Length - offset`. Then count would be the size of the chunk.

For the Transport provided SendFile, I think we could use a lower-level IHttp**Transport**SendFileFeature which doesn't have to deal with *interleaving* and *connection lifetime*.
Where you are now putting the user-level copy (under appropriate locking), we can check the lower-level feature and use that instead if available.